src/envs/pddl_env.py

```diff
@@ -0,0 +1,442 @@
+    next dimensions encode the object used to ground the operator. The
+    encoding assumes a fixed ordering over operators and objects in the
+    state.
+
+    The parameterized options are 1:1 with the STRIPS operators. They have the
+    same object parameters and no continuous parameters.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Parse the domain str.
+        self._types, self._predicates, self._strips_operators = \
+            _parse_pddl_domain(self._domain_str)
+        # The order is used for constructing actions; see class docstring.
+        self._ordered_strips_operators = sorted(self._strips_operators)
+
+    @property
+    @abc.abstractmethod
+    def _domain_str(self) -> str:
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _pddl_train_problem_generator(self) -> PDDLProblemGenerator:
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _pddl_test_problem_generator(self) -> PDDLProblemGenerator:
+        raise NotImplementedError("Override me!")
+
+    @property
+    def strips_operators(self) -> Set[STRIPSOperator]:
+        """Expose the STRIPSOperators for use by oracles."""
+        return self._strips_operators
+
+    def simulate(self, state: State, action: Action) -> State:
+        objs = list(state)
+        # Convert the state into a Set[GroundAtom].
+        ground_atoms = _state_to_ground_atoms(state)
+        # Convert the action into a _GroundSTRIPSOperator.
+        ground_op = self._action_to_ground_strips_op(action, objs)
+        # Apply the operator.
+        next_ground_atoms = utils.apply_operator(ground_op, ground_atoms)
+        # Convert back into a State.
+        next_state = _ground_atoms_to_state(next_ground_atoms, objs)
+        return next_state
+
+    def _generate_train_tasks(self) -> List[Task]:
+        return self._generate_tasks(CFG.num_train_tasks,
+                                    self._pddl_train_problem_generator,
+                                    self._train_rng)
+
+    def _generate_test_tasks(self) -> List[Task]:
+        return self._generate_tasks(CFG.num_test_tasks,
+                                    self._pddl_test_problem_generator,
+                                    self._test_rng)
+
+    def _generate_tasks(self, num_tasks: int,
+                        problem_gen: PDDLProblemGenerator,
+                        rng: np.random.Generator) -> List[Task]:
+        tasks = []
+        for pddl_problem_str in problem_gen(num_tasks, rng):
+            task = self._pddl_problem_str_to_task(pddl_problem_str)
+            tasks.append(task)
+        return tasks
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return self._predicates
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        # For now, we assume that all predicates may be included as part of the
```

Comment:
i feel like we should just fix this now. can't we just make this env-specific? alternatively we can just set self._train_tasks and self._test_tasks in `__init__` itself, and extract the goal predicates from the task goals. note that the comment in base_env.py "# These are generated lazily when..." doesn't matter here because we always have everything we need to be able to generate tasks in the init, for these pddl envs. in this case, _generate_train/test_tasks would be no-ops